src/openrct2/ride/RideConstruction.cpp

```diff
@@ -1302,132 +1265,67 @@ CoordsXYZD ride_get_entrance_or_exit_position_from_screen_position(const ScreenC
-                    if (ted.SequenceProperties[tileElement->AsTrack()->GetSequenceIndex()] & (1 << direction))
-                    {
-                        // if that side of the TrackElement supports stations, the ride entrance is valid and faces away from
-                        // the station
-                        entranceExitCoords.direction = direction_reverse(entranceExitCoords.direction);
-                        gRideEntranceExitPlaceDirection = entranceExitCoords.direction;
-                        return entranceExitCoords;
-                    }
-                } while (!(tileElement++)->IsLastForTile());
-            }
-        }
-        gRideEntranceExitPlaceDirection = INVALID_DIRECTION;
+        startDirection = mapY < 0 ? 3 : 1;
     }
-    // tracked rides excluding tower rides
     else
     {
-        // create a 2-point rectangle. The rectangle will have a width or height of 0 depending on station direction
-        CoordsXY entranceMin = { stationStart };
-        CoordsXY entranceMax = { stationStart };
-
-        // get the beginning of the station TrackElement
-        auto tileElement = ride_get_station_start_track_element(ride, gRideEntranceExitPlaceStationIndex);
-        if (tileElement == nullptr)
-        {
-            entranceExitCoords.SetNull();
-            return entranceExitCoords;
-        }
-        auto stationDirection = tileElement->GetDirection();
-        entranceExitCoords.direction = stationDirection;
-
-        auto next = entranceMax;
-        // find additional station TrackElements and extend the line of valid entrance locations
-        while (true)
+        startDirection = mapX < 0 ? 0 : 2;
+    }
+    // check all 4 directions, starting with the mouse's quadrant
+    for (uint8_t directionIncrement = 0; directionIncrement < 4; directionIncrement++)
+    {
+        entranceExitCoords.direction = (startDirection + directionIncrement) & 3;
+        // search for TrackElement one tile over, shifted in the search direction
+        auto nextLocation = entranceExitCoords;
+        nextLocation += CoordsDirectionDelta[entranceExitCoords.direction];
+        if (map_is_location_valid(nextLocation))
         {
-            entranceMax = next;
-            next -= CoordsDirectionDelta[stationDirection];
-            tileElement = map_get_first_element_at(next);
+            // iterate over every element in the tile until we find what we want
+            auto tileElement = map_get_first_element_at(nextLocation);
             if (tileElement == nullptr)
-                break;
-            bool goToNextTile = false;
-
+                continue;
             do
             {
                 if (tileElement->GetType() != TILE_ELEMENT_TYPE_TRACK)
                     continue;
-                if (tileElement->AsTrack()->GetRideIndex() != gRideEntranceExitPlaceRideIndex)
+                if (tileElement->GetBaseZ() != stationBaseZ)
```

Comment:
I think it would worth introducing ```const auto* trackElement = tileElement->AsTrack()``` here instead of calling AsTrack all over the place.

Reply:
Sounds good. Updated

Reply:
Actually, is there a reason this needs to be const auto* and not TrackElement*? There are also redundant calls in the first block of code in that function that use ->AsTrack() multiple times, so it makes sense to define a trackElement as a nullptr at the top of the function and use it for all of these instances. See the updated changes as an example.

Reply:
There is no need to specify the exact type. By using `AsTrack`, you already specify that whatever is returned has an interface to a track element. Using `auto` avoids having to change the type here when things like iterators or child classes are to be returned later on.

Reply:
Thank you for explaining. I've replaced this with auto. const doesn't seem to work here because the tileElement will change within a loop.

Reply:
> Thank you for explaining. I've replaced this with auto. const doesn't seem to work here because the tileElement will change within a loop.

My bad for the const, didn't pay enough attention. But its well advised to always use const when possible.

Reply:
Good to know. I am relatively new to C++. It's quite a bit more complicated than Python, so I appreciate the feedback.